### PR TITLE
Company Descriptions Classifier 

### PR DESCRIPTION
**Pull request by india-kerle**
---

# Description

This PR fine tunes a classification head of a bert model fine tuned on millions of job adverts to predict if a sentence is a company description or not. 

Fixes # (issue)

#4 

# Instructions for Reviewer

I've written this as a flow in `company_descriptions` folder. 

You can run the flow (not in production):

python ojd_daps_language_models/pipeline/train_model/company_descriptions/compdescs_flow.py --package-suffixes=.txt,.yaml --datastore=s3 run

Please pay special attention to ...

Although I refactored it as a flow, I actually ran it in google colab because of my metaflow dramas so I actually haven't successfully run the flow completely or tried to save the model in `save_trained_model.py` :0. Apologies if there bugs :/ 

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

**Comment by Jack-Vines** [comment]
@india-kerle at some point you've asked batch for a 6 GPU core machine (which isn't possible but batch just hangs rather than failing, which blocks all following jobs from starting, even if those have the right resources) - if you go to the [batch dashboard on AWS](https://eu-west-2.console.aws.amazon.com/batch/home?region=eu-west-2#dashboard), you can see the "runnable" jobs, with the earliest one being the one that's holding all the others up. If you terminate the ones that ask for too many resources (any of: >1 GPU, >8 CPU, >61000 memory), the rest will start successfully.

File: ojd_daps_language_models/pipeline/train_model/company_descriptions/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+| eval_steps_per_second   | 15.890000  |
+| epoch                   | 10.000000  |
+| perplexity              | 1.590000   |
+
+Test set metrics:
+
+| label                   | precision | recall   | f1-score | support |
+| ----------------------- | --------- | -------- | -------- | ------- |
+| not company description | 0.930693  | 0.959184 | 0.944724 | 98      |
+| company description     | 0.913043  | 0.857143 | 0.884211 | 49      |
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/company_descriptions/README.md:58]: this is excellent!!

File: ojd_daps_language_models/pipeline/train_model/company_descriptions/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+model = AutoModelForSequenceClassification.from_pretrained("ihk/jobbert-base-cased-compdecs")
+tokenizer = AutoTokenizer.from_pretrained("ihk/jobbert-base-cased-compdecs")
+
+comp_desc = pipeline('text-classification', model=model, tokenizer=tokenizer)
+
+job_sent = "Would you like to join a major manufacturing company?"
+comp_desc(job_sent)
+
+>> [{'label': 'LABEL_1', 'score': 0.9953641891479492}]
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/company_descriptions/README.md:31]: this ran perfectly for me fyi

File: ojd_daps_language_models/pipeline/train_model/company_descriptions/compdescs_flow.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+        date = datetime.now().strftime("%Y-%m-%d").replace("-", "")
+        self.eval_results["date"] = date
+
+        obj = io.BytesIO(json.dumps(self.eval_results).encode("utf-8"))
+        s3 = boto3.client("s3")
+        s3.upload_fileobj(
+            obj,
+            "prinz-green-jobs",
+            f"outputs/models/comp_desc_classifier/{CONFIG['model_name'].replace('-', '_')}_{date}.json",
```
Comment (lizgzil) [inline ojd_daps_language_models/pipeline/train_model/company_descriptions/compdescs_flow.py]: would be good to have the `production` flag in this output name (just to make it clear it wasnt created on a sample?)

**Review by lizgzil — COMMENTED** [review]
this looks great and such an impressive amount of labelling and model building! the scores are amazing. 

I'm also struggling to get metaflow working, so can't properly review yet.

I see you have 55 sentences tagged with 1 which are over 250 characters long, so you may up your scores even more by trying to split these into better sentence boundaries if it's possible (I guess that's for your other prinz repo PR though).

**Comment by india-kerle** [comment]
okay live updates I was able to run both the flow and the `save_trained_model.py` script !!!

The flow actually worked, it just turns out I'm not patient - when I trained the model in google colab, it took ~10 minutes but when i ran the flow (`production=False, 1 GPU, 6 CPU, 60000 memory`), it took ~1.5 hours to complete. I've slightly modified the `save_trained_model.py` script to simply save the model locally instead of also pushing the model to huggingface. 

Let me know if you're able to run it @Jack-Vines @lizgzil so I can make necessary changes and merge.

**Comment by india-kerle** [comment]
spoke with @Jack-Vines in this mornings stand up and sounds like the flow ran for him also - will squash and merge as a result!

**Pull request merged**